### ISSUE TITLE
Probe using 70% CPU

### ISSUE THREAD

**Issue by tomwilkie** [weaveworks/scope#812]
@2opremio reports the probe is using 70% CPU on the service hosts.

**Comment by 2opremio** [comment]
I restarted the probes with profiling (`--probe.http.listen :4041`)  and curled `http://localhost:4041/debug/pprof/profile` from one of the probes: http://filebin.ca/2TJeW7veYKFA/profile.gz

**Comment by tomwilkie** [comment]
![cpu](https://cloud.githubusercontent.com/assets/444037/12295134/e3de960c-b9b3-11e5-828b-b638f760b738.png)

**Comment by tomwilkie** [comment]
At a cursory analysis, the probes are spending:
- 30% of the CPU reading the /proc/<pid>/net/tcp files
- maybe <10% of cpu is spent on GC (<3 of 30s) according to that trace
- 10% on copying in the report structure

I suspect fixing the copying will also fix the GC, but I can’t think of anything we can do to ‘fix’ the proc cpu usage, as this has already been heavily optimised (we only read the tcp files for net namespaces we haven't seen before).

I guess we can decrease the polling frequency / proc spy interval, and maybe even come up with some rate limiting scheme (ie only read 10/s, which would get us ~constant cpu usage, and at larger scales would means the UI updates less frequently).

**Comment by tomwilkie** [comment]
@2opremio tried that suggestion (`./scope -probe.spy.interval=3s`) and confirmed CPU dropped to 20%, as expected.

**Comment by 2opremio** [comment]
As suggested by @tomwilkie , decreasing the frequency to

**Comment by 2opremio** [comment]
The situation improves considerably after #898 . Now the probes (2ed3968) are using 30-40% CPU with peaks of 50%

Here go two probe profiles I have obtained:

[pprof.localhost_4041.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/116782/pprof.localhost_4041.samples.cpu.001.pb.gz)
![probe](https://cloud.githubusercontent.com/assets/2362916/12804092/bb9a7dce-cae7-11e5-8fea-71b404fc9cef.png)

[pprof.localhost_4041.samples.cpu.002.pb.gz](https://github.com/weaveworks/scope/files/116783/pprof.localhost_4041.samples.cpu.002.pb.gz)
![probe2](https://cloud.githubusercontent.com/assets/2362916/12804098/c8a95a26-cae7-11e5-9f69-d2a54a57ef7e.png)


* The probes are spending a long time encoding reports. A non-negligible amount of time (10% in one of the profiles) is spent sending gob types, which can be brought to 0 through any of the proposed solutions from https://github.com/weaveworks/scope/issues/854#issuecomment-179560048

* Although the Reads are not the bottleneck anymore thanks to #898, the `stat()`s of `/proc/*/fd/*` are consuming a non negligible amount of time (10% and 35% in the reports above). So we should maybe reconsider  #814, which I should had tested with more open files: `sudo bash -c "echo /proc/*/fd/* | wc -w"` prints numbers up to ~20K in the service machines, whereas my own machine only has 1K.

* Kubernetes support (which is enabled in the probes) takes around 10% of the time in the second profile (I think this is the cause of the 10% CPU peaks), almost exclusively matching pods:

```
$ go tool pprof  ~/home-vagrant/scope-tested pprof.localhost_4041.samples.cpu.002.pb.gz 
Entering interactive mode (type "help" for commands)
(pprof) list podTopology.func1
Total: 11.78s
ROUTINE ======================== github.com/weaveworks/scope/probe/kubernetes.(*Reporter).podTopology.func1 in /go/src/github.com/weaveworks/scope/probe/kubernetes/reporter.go
      10ms      1.16s (flat, cum)  9.85% of Total
         .          .     51:}
         .          .     52:
         .          .     53:func (r *Reporter) podTopology(services []Service) (report.Topology, report.Topology, error) {
         .          .     54:   pods, containers := report.MakeTopology(), report.MakeTopology()
         .          .     55:   err := r.client.WalkPods(func(p Pod) error {
      10ms       10ms     56:           for _, service := range services {
         .      1.10s     57:                   if service.Selector().Matches(p.Labels()) {
         .          .     58:                           p.AddServiceID(service.ID())
         .          .     59:                   }
         .          .     60:           }
         .          .     61:           nodeID := report.MakePodNodeID(p.Namespace(), p.Name())
         .       40ms     62:           pods = pods.AddNode(nodeID, p.GetNode())
         .          .     63:
         .          .     64:           container := report.MakeNodeWith(map[string]string{
         .          .     65:                   PodID:     p.ID(),
         .          .     66:                   Namespace: p.Namespace(),
         .       10ms     67:           }).WithParents(report.EmptySets.Add(report.Pod, report.MakeStringSet(nodeID)))
         .          .     68:           for _, containerID := range p.ContainerIDs() {
         .          .     69:                   containers.AddNode(report.MakeContainerNodeID(containerID), container)
         .          .     70:           }
         .          .     71:           return nil
         .          .     72:   })
(pprof) list Selector
Total: 11.78s
ROUTINE ======================== github.com/weaveworks/scope/probe/kubernetes.(*service).Selector in /go/src/github.com/weaveworks/scope/probe/kubernetes/service.go
         0      1.04s (flat, cum)  8.83% of Total
         .          .     47:
         .          .     48:func (s *service) Selector() labels.Selector {
         .          .     49:   if s.Spec.Selector == nil {
         .          .     50:           return labels.Nothing()
         .          .     51:   }
         .      1.04s     52:   return labels.SelectorFromSet(labels.Set(s.Spec.Selector))
         .          .     53:}
         .          .     54:
         .          .     55:func (s *service) GetNode() report.Node {
         .          .     56:   return report.MakeNodeWith(map[string]string{
         .          .     57:           ServiceID:      s.ID(),

````

@paulbellamy Any ideas on how improve this? [SelectroFromSet](https://github.com/kubernetes/kubernetes/blob/master/pkg/labels/selector.go#L774) seems to be prohibitively expensive, mostly due to the regular expression matching (maybe we should fix that):

```
(pprof) list SelectorFromSet
Total: 11.78s
ROUTINE ======================== github.com/weaveworks/scope/vendor/k8s.io/kubernetes/pkg/labels.SelectorFromSet in /go/src/github.com/weaveworks/scope/vendor/k8s.io/kubernetes/pkg/labels/selector.go
      30ms      1.04s (flat, cum)  8.83% of Total
         .          .    719:func SelectorFromSet(ls Set) Selector {
         .          .    720:   if ls == nil {
         .          .    721:           return LabelSelector{}
         .          .    722:   }
         .          .    723:   var requirements []Requirement
      20ms       30ms    724:   for label, value := range ls {
         .      890ms    725:           if r, err := NewRequirement(label, EqualsOperator, sets.NewString(value)); err != nil {
         .          .    726:                   //TODO: double check errors when input comes from serialization?
         .          .    727:                   return LabelSelector{}
         .          .    728:           } else {
      10ms       80ms    729:                   requirements = append(requirements, *r)
         .          .    730:           }
         .          .    731:   }
         .       40ms    732:   return LabelSelector(requirements)
         .          .    733:}
(pprof) list NewRequirement
Total: 11.78s
ROUTINE ======================== github.com/weaveworks/scope/vendor/k8s.io/kubernetes/pkg/labels.NewRequirement in /go/src/github.com/weaveworks/scope/vendor/k8s.io/kubernetes/pkg/labels/selector.go
      20ms      660ms (flat, cum)  5.60% of Total
         .          .    105:// (5) The key is invalid due to its length, or sequence
         .          .    106://     of characters. See validateLabelKey for more details.
         .          .    107://
         .          .    108:// The empty string is a valid value in the input values set.
         .          .    109:func NewRequirement(key string, op Operator, vals sets.String) (*Requirement, error) {
      10ms      260ms    110:   if err := validateLabelKey(key); err != nil {
         .          .    111:           return nil, err
         .          .    112:   }
         .          .    113:   switch op {
         .          .    114:   case InOperator, NotInOperator:
         .          .    115:           if len(vals) == 0 {
         .          .    116:                   return nil, fmt.Errorf("for 'in', 'notin' operators, values set can't be empty")
         .          .    117:           }
      10ms       10ms    118:   case EqualsOperator, DoubleEqualsOperator, NotEqualsOperator:
         .          .    119:           if len(vals) != 1 {
         .          .    120:                   return nil, fmt.Errorf("exact-match compatibility requires one single value")
         .          .    121:           }
         .       10ms    122:   case ExistsOperator, DoesNotExistOperator:
         .          .    123:           if len(vals) != 0 {
         .          .    124:                   return nil, fmt.Errorf("values set must be empty for exists and does not exist")
         .          .    125:           }
         .          .    126:   default:
         .          .    127:           return nil, fmt.Errorf("operator '%v' is not recognized", op)
         .          .    128:   }
         .          .    129:
         .       20ms    130:   for v := range vals {
         .      320ms    131:           if err := validateLabelValue(v); err != nil {
         .          .    132:                   return nil, err
         .          .    133:           }
         .          .    134:   }
         .       40ms    135:   return &Requirement{key: key, operator: op, strValues: vals}, nil
         .          .    136:}
         .          .    137:
         .          .    138:// Matches returns true if the Requirement matches the input Labels.
         .          .    139:// There is a match in the following cases:
         .          .    140:// (1) The operator is Exists and Labels has the Requirement's key.
(pprof) 
```

**Comment by tomwilkie** [comment]
We made the gob encoding once-per-report a while back (it used to be
onces-per-report-per-app).  That was a significant improvement.  If we want
to do gob-over-websocket, we'll have to go back to a separate encoder per
app (as different app connections will have different lifecycles), which I
worry will be more of a regression in CPU usage.

How many probes are there?  How many apps?  I suspect we might just be
hitting the O(n*m) nature of this architecture.  Going for 2 apps and N
probes should be an improvement?

Other things you highlight also look good:
- rate limit reading of FDs, although due to the race with
/proc/PID/net/tcp this will have to look like a periodic reading of
namespaces, with some dynamic-feedback based on the number of fds in a
namespace, which might be complicated
- kubernetes optimisations

Thanks for working on this fons!

**Comment by 2opremio** [comment]
The rate limiting helps quite a bit:

![probe_rate_limit](https://cloud.githubusercontent.com/assets/2362916/12824044/265d563e-cb67-11e5-91e2-59069067b740.png)


Now the profile is dominated by garbage collection and the kubernetes problem.

**Comment by 2opremio** [comment]
After cherry-picking @paul's https://github.com/weaveworks/scope/tree/k8s-efficiency on top of the rate limiting, the CPU consumption is between 25% and 35%:

![probe_rate_limit_k8s](https://cloud.githubusercontent.com/assets/2362916/12824505/6635a25a-cb69-11e5-9184-ef695f88ff69.png)

Now the profile is completely dominated by encoding, copying and garbage collection. I am going to work on the generated json encoding, clean and get #912 in and then see how we can improve the garbage collection dominance if it's still there.

**Comment by 2opremio** [comment]
After the codec improvements, the probes still consume ~20% CPU in the service.

[pprof.localhost:4041.samples.cpu.001.pb.gz](https://github.com/weaveworks/scope/files/136685/pprof.localhost.4041.samples.cpu.001.pb.gz)


![probe](https://cloud.githubusercontent.com/assets/2362916/13153068/0dfc855c-d66a-11e5-9b00-ea0bcb83979b.png)

The main two places where to improve are:

* Report copying. It consumes a lot of CPU and generates a lot of objects, putting a lot pressure in the GC
* The codec. Yes, it still takes 17% of the time, but I think the solution here is to simply reduce the size of the reports or their frequency. I have created https://github.com/weaveworks/scope/issues/985

**Comment by 2opremio** [comment]
@tomwilkie @paulbellamy @peterbourgon Any ideas on how to reduce or optimize the copying?

**Comment by peterbourgon** [comment]
We can save a ton of allocs and CPU time by leveraging [sync.Pool](https://golang.org/pkg/sync/#Pool). For sure we can use pools of []byte or *bytes.Buffer when we serialize reports (and deserialize on the other end). That's easy and may get us some quick wins.

Beyond that, it'd be great to use pools for the reports themselves, i.e. don't allocate a new report and fill it, but fetch one from the pool, reset it, and write into it. But, we'd need implement a Reset method, and I'm not sure how simple that would be given all of our nested maps. 

Other thoughts?

**Comment by paulbellamy** [comment]
It's a bit unclear to me why copying is taking that much CPU/making that much garbage. Copy should be *almost* a no-op. Unless there is stuff left which is being copied? It looks a bit from the graph like Metrics?

Alternatively we could ditch the whole immutability/copying thing and rethink how the pipeline works. Go doesn't exactly make working with immutable datastructures easy/elegant anyway.

**Comment by peterbourgon** [comment]
> Alternatively we could ditch the whole immutability/copying thing and rethink how the pipeline works. Go doesn't exactly make working with immutable datastructures easy/elegant anyway.

:+1: to this. The current report structure and pipeline architecture was designed for ease of use and expediency. if it's no longer easy to use (i.e. maintain and mutate with changing product requirements) and it's a performance bottleneck with no obvious fix, it's absolutely appropriate to revisit the overall design.

**Comment by 2opremio** [comment]
> We can save a ton of allocs and CPU time by leveraging sync.Pool. For sure we can use pools of []byte or *bytes.Buffer when we serialize reports (and deserialize on the other end).

AFAIK the new codec library already uses pools.

However, I believe that the fromIntermediate/toIntermediate methods used for custom serialization (e.g. the ones needed for mndrix/ps like LatestMmap) may be expensive expensive

> Beyond that, it'd be great to use pools for the reports themselves

This sounds more like it!

**Comment by peterbourgon** [comment]
> This sounds more like it!

Unfortunately, after a bit of research and discussion with the community, I don't think the current report structure is going to lend itself to Pooling very well. In the end it's a collection of maps, and maps only 'reset' in linear time, i.e.

```go
for _, topo := range report.Topologies() {
    for k := range topo {
        delete(topo, k)
    }
}
```

It's worth a profile, of course, but the cost of reset increases as the map gets bigger, making it perversely incentivized against cost of GC. My intuition is that if we want to Pool reports, we need to get rid of most if not all of the maps, in favor of slices or something else — IMO worth considering together with @paulbellamy's bag-of-nodes approach to the pipelines.

**Comment by 2opremio** [comment]
Actually, after having a closer look at the profile, the 17% of time in the codec doesn't come from Encoding but Decoding, so it can't come from the reports being sent to the app. 10% of that time is spent reading http bodies, so the decoding itself os only 7%. 

Docker/kubernetes info? It's hard to tell because apparently it's being done in a separate goroutine. I'll investigate but we may be pooling something way too often.

If needed, to accelerate the decoding of that information we can always generate `Selfer`s for those types in the vendor directory.

**Comment by 2opremio** [comment]
Also, to get a better sense of where we can reduce garbage generation, I just found out that pprof lets you obtain heap profiles with `--alloc_objects`, which will tell you how many objects have been allocated overall (not just in use). I will also try that.

**Comment by 2opremio** [comment]
Yep, it's docker:

```
(pprof) list StartGatheringStats
Total: 6.87s
ROUTINE ======================== github.com/weaveworks/scope/probe/docker.(*container).StartGatheringStats.func1 in /go/src/github.com/weaveworks/scope/probe/docker/container.go
         0      1.23s (flat, cum) 17.90% of Total
         .          .    202:			c.statsConn = nil
         .          .    203:		}()
         .          .    204:
         .          .    205:		var stats docker.Stats
         .          .    206:		decoder := codec.NewDecoder(resp.Body, &codec.JsonHandle{})
         .      1.23s    207:		for err := decoder.Decode(&stats); err != io.EOF; err = decoder.Decode(&stats) {
         .          .    208:			if err != nil {
         .          .    209:				log.Errorf("docker container: error reading event, did container stop? %v", err)
         .          .    210:				return
         .          .    211:			}
         .          .    212:
```

**Comment by paulbellamy** [comment]
go-dockerclient (and docker API) has a [streaming stats thing](https://godoc.org/github.com/fsouza/go-dockerclient#Client.Stats) that *might* reduce this.

**Comment by tomwilkie** [comment]
Nah, there code is basically identical.

**Comment by 2opremio** [comment]
Here's the `--alloc_objects` heap profile for a probe (d7442788548c53971aaccdb681a274db31ba57a3) running in the service:

[pprof.localhost:4041.alloc_objects.alloc_space.001.pb.gz](https://github.com/weaveworks/scope/files/138451/pprof.localhost.4041.alloc_objects.alloc_space.001.pb.gz)


![probe_obj_allocs](https://cloud.githubusercontent.com/assets/2362916/13184261/b00d0b28-d732-11e5-821c-40759e11fb9d.png)

**Comment by 2opremio** [comment]
Final measurement for a while, (after #1000). The probes are running stably under 20% (whilst kubelet 1.1 is running >=70% in all nodes) and the garbage generated is _tolerable_:

CPU profile:

[pprof.localhost:4041.samples.cpu.002.pb.gz](https://github.com/weaveworks/scope/files/140900/pprof.localhost.4041.samples.cpu.002.pb.gz)


![probe](https://cloud.githubusercontent.com/assets/2362916/13223647/076d0052-d97d-11e5-8296-d24b64cbdc45.png)

Objects allocations profile:

[pprof.localhost:4041.alloc_objects.alloc_space.001.pb.gz](https://github.com/weaveworks/scope/files/140899/pprof.localhost.4041.alloc_objects.alloc_space.001.pb.gz)



![app_object_alloc](https://cloud.githubusercontent.com/assets/2362916/13223681/357b0b88-d97d-11e5-957d-0c544a50ee60.png)

**Comment by 2opremio** [comment]
After making another measurement and looking at the syscalls (`-focus 'Syscall$'`), I can see that the Syscall load is shared between the connection walker (probe/endpoint/procspy/proc_linux.go) and the process walker (probe/process/walker_linux.go), which after #729 also inspects the fd directories:

[pprof.localhost:4041.samples.cpu.003.pb.gz](https://github.com/weaveworks/scope/files/142579/pprof.localhost.4041.samples.cpu.003.pb.gz)

![probe_syscall](https://cloud.githubusercontent.com/assets/2362916/13252083/522561ac-da2c-11e5-9e3b-b5b6b3a1a99c.png)

**Comment by 2opremio** [comment]
Closing in favor of #1010 and #1011
